### ISSUE TITLE
Issues when converting from IFC

### ISSUE THREAD

**Issue by HarpaceasDevelopment** [xeokit/xeokit-convert#86]
From Command Line this is the result I am getting: 
C:\Users\me\Desktop\HarpaceasBIMViewer\next-harpa-bim-viewer\node_modules\@xeokit\xeokit-convert>node convert2xkt.js -s <path to get file Duplex.ifc> -o test.xkt -l
[convert2xkt] Running convert2xkt v1.1.4...
[convert2xkt] Reading input file: C:\Users\me\Desktop\PonteArno_CV2.ifc
[convert2xkt] Input file size: 10024.28 kB
Error: Error: unsupported source format: "ifc".

I tried with several IFC files but I get the same result.

from Javascript (node.js) I am trying to do the same with following code:
`import {convert2xkt} from '@xeokit/xeokit-convert/dist/convert2xkt.cjs.js'

convert2xkt({
    source: "path to get file Duplex.ifc",
    output: "outputpath.xkt",
    log: (msg) => {
        console.log(msg)
    }
}).then(() => {
    console.log("Converted.");
}, (errMsg) => {
    console.error("Conversion failed: " + errMsg)
});`

but I get this output:

[convert2xkt] Reading input file: ... Duplex.ifc
[convert2xkt] Input file size: 2366.05 kB
[convert2xkt] Using parser: parseIFCIntoXKTModel
Conversion failed: TypeError: Cannot read properties of undefined (reading 'IfcAPI')

Does these logs help you undertstand which could be the problem?

**Comment by xeolabs** [comment]
Hi, 

for the first issue, it seems that you're not specifying an IFC input file on the command line - your command is:

````
node convert2xkt.js -s -o test.xkt -l
````

where it should be something like: 

````
node convert2xkt.js -s myFile.ifc -o test.xkt -l
````

Other than that, I'm not sure what the problem could be here - I haven't seen this error with tests so far.

Maybe try on Ubuntu?

Or perhaps try a different pipeline, as described here: https://www.notion.so/Converting-IFC-Models-to-XKT-using-3rd-Party-Open-Source-Tools-c373e48bc4094ff5b6e5c5700ff580ee

Also, share the file if you like and I can try on my end.

**Comment by HarpaceasDevelopment** [comment]
Thank you very much for your quick response!
For the command line I omitted by mistake the input file in this issue description, this is what I actually got:
```
C:\Users\manigrasso\Desktop\HarpaceasBIMViewer\next-harpa-bim-viewer\node_modules\@xeokit\xeokit-convert>node convert2xkt.js -f ifc -s C:\Users\manigrasso\Desktop\Sensori_4_UpdatePset_SHARED.ifc  -o test.xkt -l
[convert2xkt] Running convert2xkt v1.1.4...
[convert2xkt] Reading input file: C:\Users\manigrasso\Desktop\Sensori_4_UpdatePset_SHARED.ifc
[convert2xkt] Input file size: 1658.76 kB
Error: Error: unsupported source format: "ifc".
```

this is the same result I get with every ifc file using xeokit-convert.

I am managing to get it to work with other pipelines as suggested by you ( uploading also the latest ifcConvert in "ifc-converter", because there's not the latest the one in that package) but it  works well only with Surface geometry (SG) not with CV2.

example CV2 model converted with alternative pipeline : 
![immagine](https://user-images.githubusercontent.com/116150912/204843912-c8796dda-021f-49d8-b7a6-1b877556b8b4.png)

as you can see despite I am using the latest IfcConvert it is not creating the expected geometry, that's why I would like to make the xeokit-convert to work.

If I understand the reason of this behaviour I will update you.

P.s. Actually I'm on Windows for testing, I will try also on Ubuntu.

Thank you again.

**Comment by xeolabs** [comment]
Thanks!

Note that the convert2xkt function is using [web-ifc](https://github.com/IFCjs/web-ifc) internally, to parse the geometry and metadata from the IFC,

So the missing CV2 geometry is because either 1) its not supported by the geometry engine in web-ifc, or 2) we're just not using some feature of web-ifc yet. I think it would be (1).

The convert2xkt direct IFC->XKT conversion is not our "enterprise-recommended" pipeline as such - it's more for development and experimentation. 

This means that pipeline can be lacking  some of the newer features that we support in our more enterprise-ready [ifc2gltf](https://www.notion.so/Converting-IFC-to-XKT-using-ifc2gltf-a2e0005d00dc4f22b648f1237bc3245d) pipeline. 

For example, in that pipeline we parse property sets more completely, which I think might help with the showing those missing properties.   I think we may have some more property set types to convert yet, but we convert more.

So, the best conversion technique IMO would be:

- use ifc2gltf to create a .glb file containing geometry and a .json file containing metadata (elements and property sets)
- then convert the .glb and .json into .xkt using convert2xkt (this time using it's .glb+.json -> .xkt mode)

Then hopefully the .xkt should contain the property sets and geometry you need. If the geometry is missing, that probably means the geometry engine in ifc2gltf doesn't handle it yet. If property sets are missing, possibly that ifc2gltf are not handling those properties. In both bases, we should be able to fix ifc2gltf.

I think you now have access to ifc2gltf, so from here let's see if it converts the geometry and property sets you need..then perhaps we have something to add or fix there, let's see.

**Comment by HarpaceasDevelopment** [comment]
Hello,
Thank you very much for the very clear answer, I appreciate.

I managed to convert an Ifc file directly into Xkt with all the related data.
I think the problem was due to the fact that using next.js there are differences in the implementation, this is the javascript code I use to make the conversion work:

```
convert2xkt({
              WebIFC : WebIFC,
              source: asset, 
              output: outputFile,
              log: (msg) => {
                  console.log(msg)
    },
}
```
 

Debugging in the “convert2xkt” I found that ‘WebIFC’ parameter was undefined, so I have to import it and pass it as a parameter.
I have to say that conversion with simple models works very nice.

But I have some models in which the conversion gives the following exceptions:

1) ‘Conversion failed: RuntimeError: memory access out of bounds’
2) '[24276:000002105E569B70]  1535309 ms: Mark-sweep 4052.7 (4140.7) -> 4038.5 (4142.7) MB, 1304.3 / 0.0 ms  (average mu = 0.269, current mu = 0.230) allocation failure scavenge might not succeed
[24276:000002105E569B70]  1537872 ms: Mark-sweep 4054.7 (4142.9) -> 4040.6 (4144.9) MB, 2497.8 / 0.0 ms  (average mu = 0.122, current mu = 0.026) allocation failure scavenge might not succeed FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
1: 00007FF7C30A815F v8::internal::CodeObjectRegistry::~CodeObjectRegistry+114079
 2: 00007FF7C30354C6 DSA_meth_get_flags+65542
 3: 00007FF7C303637D node::OnFatalError+301
 4: 00007FF7C396BA0E v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF7C3955FED v8::SharedArrayBuffer::Externalize+781
 6: 00007FF7C37F93BC v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF7C3806069 v8::internal::Heap::PublishPendingAllocations+1129
 8: 00007FF7C380303A v8::internal::Heap::PageFlagsAreConsistent+2842
 9: 00007FF7C37F5C99 v8::internal::Heap::CollectGarbage+2137
10: 00007FF7C37F3E50 v8::internal::Heap::AllocateExternalBackingStore+2000
11: 00007FF7C38189D6 v8::internal::Factory::NewFillerObject+214
12: 00007FF7C354AEA5 v8::internal::DateCache::Weekday+1797
13: 00007FF7C39F9701 v8::internal::SetupIsolateDelegate::SetupHeap+494417
14: 00007FF7C39FA6D3 v8::internal::SetupIsolateDelegate::SetupHeap+498467
15: 000001747832307E’



So as suggested [here ](https://xeokit.github.io/xeokit-convert/docs/):

‘Direct IFC conversion is an alpha status feature, since it depends on [web-ifc](https://github.com/tomvandig/web-ifc), a 3rd-party library, which is also alpha at this time. As such, some IFC models may not convert properly. If this is the case for your models, consider using our [standard conversion setup](https://www.notion.so/xeokit/Converting-IFC-Models-using-3rd-Party-Open-Source-Tools-c373e48bc4094ff5b6e5c5700ff580ee) until issues have been resolved.’

So I managed to get the geometry and hierarchy metadata from other pipelines (I got better results with [this], only geometric issues I found are due to Coordination View 2.0(https://www.notion.so/Converting-IFC-Models-to-XKT-using-3rd-Party-Open-Source-Tools-c373e48bc4094ff5b6e5c5700ff580ee), but as you know the property sets are missing, at first I thought to modify _xeokit-metadata_ source code in order to add them, but I don't know the exact schema expected for the correct conversion in xkt, and I don't know if this would be desired or not.
if yes is possible to get it?

By the way now that you confirmed me that ifc2gltf also manages the property sets, I will manage to get the access to it as soon as possible.

Thank you so much again!
